plugins/net_plugin/net_plugin.cpp

```diff
@@ -2075,8 +2137,43 @@ namespace eosio {
    }
 
    //------------------------------------------------------------------------
+   // thread safe
 
+   void unlinkable_block_state_cache::add_unlinkable_block( signed_block_ptr b, const block_id_type& id ) {
+      std::unique_lock g(unlinkable_blk_state_mtx);
+      unlinkable_blk_state.insert( {id, std::move(b)} ); // does not insert if already there
+      if (unlinkable_blk_state.size() > max_unlinkable_cache_size) {
+         auto& index = unlinkable_blk_state.get<by_timestamp>();
+         auto begin = index.begin();
+         block_id_type rm_block_id = begin->id;
+         index.erase( begin );
+         g.unlock();
+         // rm_block since we are no longer tracking this not applied block, allowing it to flow back in if needed
+         dispatch.rm_block(rm_block_id);
```

Comment:
Thanks a lot for making the change, Kevin. This line is the only location where `dispatch` is used. Instead of creating a dependency, why not have this function returning `std::optional<block_id_type>` as in my original suggestion? That way `unlinkable_block_state_cache` doesn't know about `dispatch_manager`. And you won't need the `g.unlock()`.

Also, why not have the functions inline in the class? It is all in the `.cpp` file anyways.